### PR TITLE
fix #140, add matrix.map

### LINKED ISSUE #140: in-place operations occur in matrix.js
e.g.
```javascript
n = require('numbers')
var m1 = [[1,2,3],[4,5,6],[7,8,9]];
var m2 = n.matrix.scalar(m1, 10);
```

this function alters `m1` as well.

I think it would be good to have an option for having both in-place operations and ... not in-place. Julia (http://julialang.org) uses the `!` character to represent in-place operations, but we are not so lucky to have no access to this character. any suggestions?

### PR DESCRIPTION
**Pull request by SaintDako**
#140 wasn't broken, per se, but an in-place operation is deceiving. as mentioned in that issue, it would be nice to have a separate function for an in-place operation, in case it's needed. perhaps a suffix of `IP` would be suitable.

`matrix.map` takes a matrix (or vector) and applies a function to every element to a copy of the matrix (not in-place!). the purpose behind this function is to imitate how some of Numpy's functions operate element-by-element on Numpy arrays. so instead of writing `matrix.sin`, `matrix.cos`... etc. the user can just pass any function to `matrix.map`.

the catch: the function passed can only take one parameter, so it would be nice (and pretty simple, methinks) to create something that would behave like so:

```
f = function(x, a, b) {
  return a*x + b;
}
var m = [[1,2,3],[4,5,6]];

var m2 = matrix.mapNEW(m, f, a, b);
```

so that any additional parameters passed to `matrix.mapNEW` (name is a work in progress) are passed to `f` too.

### CODE REVIEW

**Comment by philipithomas** [comment]
Shouldn't this be called a "map" function?

**Comment by SaintDako** [comment]
Yes you're correct. I was struggling to come up with a name. Will fix in a
bit

**Comment by SaintDako** [comment]
@KartikTalwar mind reviewing this? should only take a minute :hamburger:

File: lib/numbers/matrix.js [hunk 1]
```diff
@@ -1288,3 +1289,27 @@ matrix.isStrictlyColumnDD = function (M) {
...
+  // M is n-by-m (n rows, m columns)
+  var n = M.length,
+      m = M[0].length,
+      i, j;
+
+  var res = matrix.deepCopy(M);
+
+  for (i=0; i<n; i++) {
+    for (j=0; j<m; j++) {
+      res[i][j] = f(M[i][j]);
```
Comment (KartikTalwar) [inline lib/numbers/matrix.js]: I think `.apply` would be better instead of calling the function but it's not that big of a deal.

**Comment by KartikTalwar** [comment]
Mostly lgtm :hamburger:

**Comment by SaintDako** [comment]
looking at some [benchmarks](https://jsperf.com/function-calls-direct-vs-apply-vs-call-vs-bind/6), doing a direct call is faster. I think doing this for a single argument is okay, but if we want to generalize it we can use `apply`.

**Comment by KartikTalwar** [comment]
Cool, feel free to merge

**Comment by SaintDako** [comment]
anyone have a suggestion for the name of the second version of `matrix.map`?

**Pull request merged**